cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll

```diff
@@ -734,14 +736,32 @@ class TranslatedTryStmt extends TranslatedStmt {
     // of the `try`, because the exception successor of the `try` itself is
     // the first catch clause.
     handler = this.getHandler(stmt.getNumberOfCatchClauses() - 1) and
-    result = this.getParent().getExceptionSuccessorInstruction(kind)
+    exists(boolean isSEH |
+      stmt instanceof MicrosoftTryStmt and isSEH = true
+      or
+      stmt instanceof TryStmt and isSEH = false
```

Comment:
Hmm... I'm not totally convinced this is correct. Essentially, this is where we tell outer `try`/`catch` statements what kind of exception has been thrown. I get that there's probably a very strong correlation between whether it's a C++ exception or a SEH exception, and whether it was a `try`/`catch` or a `__try`/`__catch` (at least in correct code), but we shouldn't rely on this!

Instead, we need a way to "remember" what the original exception kind was and propagate that up. Looking at calls to `getNextHandler` I only see two calls: https://github.com/search?q=repo%3Agithub%2Fcodeql%20getNextHandler&type=code

For the call in `TranslatedCatchByTypeHandler` we know that the `kind` is an `ExceptionKind` and we can get the Boolean (which describes whether or not it's SEH) from there.

For the call in `TranslatedDestructorsAfterHandler` we need a bit more work. `TranslatedDestructorsAfterHandler` is also only produced in 1 place: https://github.com/github/codeql/blob/d0f73acc254962d5ea0f290fd3c47074eced9de4/cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll#L951

Notice that this predicate is only called in 1 place: on the line right above the `getNextHandler` call in `TranslatedCatchByTypeHandler` from before.

So we can also get the Boolean in that case. We then need to store that Boolean somewhere. To do that, we need to add a `Boolean` column to `TTranslatedDestructorsAfterHandler`.

I know this is quite a lot of explanation 😅 Let me know if you need more directions here!

Reply:
Per an offline discussion, we have currently decided to explicitly avoid mix and match exceptions, but after this PR is approved we want to open an issue to investigate this assumption in the future.